lib/ofdm_coarse_frequency_correction_vcvc_impl.cc

```diff
@@ -0,0 +1,162 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2017 Moritz Luca Schmid, Communications Engineering Lab (CEL) / Karlsruhe Institute of Technology (KIT).
+ *
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3, or (at your option)
+ * any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <gnuradio/io_signature.h>
+#include "ofdm_coarse_frequency_correction_vcvc_impl.h"
+#include <math.h>
+
+namespace gr {
+  namespace dab {
+
+    ofdm_coarse_frequency_correction_vcvc::sptr
+    ofdm_coarse_frequency_correction_vcvc::make(int fft_length, int num_carriers, int cyclic_prefix_length)
+    {
+      return gnuradio::get_initial_sptr
+        (new ofdm_coarse_frequency_correction_vcvc_impl(fft_length, num_carriers, cyclic_prefix_length));
+    }
+
+    /*
+     * The private constructor
+     */
+    ofdm_coarse_frequency_correction_vcvc_impl::ofdm_coarse_frequency_correction_vcvc_impl(int fft_length, int num_carriers, int cyclic_prefix_length)
+      : gr::sync_block("ofdm_coarse_frequency_correction_vcvc",
+              gr::io_signature::make(1, 1, fft_length*sizeof(gr_complex)),
+              gr::io_signature::make(1, 1, num_carriers*sizeof(gr_complex))),
+        d_fft_length(fft_length),
+        d_num_carriers(num_carriers),
+        d_cyclic_prefix_length(cyclic_prefix_length)
+    {
+      d_freq_offset = 0;
+      d_snr = 0;
+    }
+
+    /*
+     * Our virtual destructor.
+     */
+    ofdm_coarse_frequency_correction_vcvc_impl::~ofdm_coarse_frequency_correction_vcvc_impl()
+    {
+    }
+
+    /*! Energy measurement over the num_carriers sub_carriers + the central carrier.
+     * The energy gets a maximum when the calculation window and the occupied carriers are congruent.
+     * Fine frequency synchronization in the range of one sub-carrier spacing is already done.
+     */
+    void
+    ofdm_coarse_frequency_correction_vcvc_impl::measure_energy(const gr_complex *symbol)
+    {
+      unsigned int i, index;
+      float energy = 0, max = 0;
+      // first energy measurement is processed completely
+      for (i=0; i<d_num_carriers+1; i++) {
+        if (i != d_num_carriers/2)
+          energy+=std::real(symbol[i]*conj(symbol[i]));
```

Comment:
VOLKize the shit out of this: this is just mag_squared, there's a VOLK kernel for that.